ktor-server/jvm/src/io/ktor/server/application/plugins/api/RelativePluginContexts.kt

```diff
@@ -0,0 +1,171 @@
+
+/**
+ * A [PluginContext] context that allows inserting [currentPlugin] actions relatively (before/after) to [otherPlugins].
+ **/
+public abstract class RelativePluginContext(
+    private val currentPlugin: ServerPlugin<*>,
+    private val otherPlugins: List<ServerPlugin<*>>
+) : PluginContext {
+    private fun <T : Any> sortedPhases(
+        interceptions: List<Interception<T>>,
+        pipeline: Pipeline<*, ApplicationCall>,
+        otherPlugin: ServerPlugin<*>
+    ): List<PipelinePhase> =
+        interceptions
+            .map { it.phase }
+            .sortedBy {
+                if (!pipeline.items.contains(it)) {
+                    throw PluginNotInstalledException(otherPlugin.name)
+                }
+
+                pipeline.items.indexOf(it)
+            }
+
+    protected abstract fun selectPhase(phases: List<PipelinePhase>): PipelinePhase?
+
+    protected abstract fun insertPhase(
+        pipeline: Pipeline<*, ApplicationCall>,
+        relativePhase: PipelinePhase,
+        newPhase: PipelinePhase
+    )
+
+    private fun <T : Any, ContextT : CallHandlingContext> insertToPhaseRelativelyWithMessage(
+        currentInterceptions: MutableList<Interception<T>>,
+        otherInterceptionsList: List<MutableList<Interception<T>>>,
+        contextInit: (PipelineContext<T, ApplicationCall>) -> ContextT,
+        block: suspend ContextT.(ApplicationCall, Any) -> Unit
+    ) {
+        val currentPhase = currentPlugin.newPhase()
+
+        currentInterceptions.add(
+            Interception(
+                currentPhase,
+                action = { pipeline ->
+                    for (i in otherPlugins.indices) {
+                        val otherPlugin = otherPlugins[i]
+                        val otherInterceptions = otherInterceptionsList[i]
+
+                        val otherPhases = sortedPhases(otherInterceptions, pipeline, otherPlugin)
+                        selectPhase(otherPhases)?.let { lastDependentPhase ->
+                            insertPhase(pipeline, lastDependentPhase, currentPhase)
+                        }
+                    }
+
+                    pipeline.intercept(currentPhase) {
+                        contextInit(this).block(call, subject)
+                    }
+                }
+            )
+        )
+    }
+
+    private fun <T : Any, ContextT : CallHandlingContext> insertToPhaseRelatively(
+        currentInterceptions: MutableList<Interception<T>>,
+        otherInterceptions: List<MutableList<Interception<T>>>,
+        contextInit: (PipelineContext<T, ApplicationCall>) -> ContextT,
+        block: suspend ContextT.(ApplicationCall) -> Unit
+    ) = insertToPhaseRelativelyWithMessage(currentInterceptions, otherInterceptions, contextInit) { call, _ ->
+        block(call)
+    }
+
+    override val onCall: OnCall = object : OnCall {
+        override operator fun invoke(block: suspend CallContext.(ApplicationCall) -> Unit) {
+            insertToPhaseRelatively(
```

Comment:
Are we going to sort interceptors on every call?

Reply:
No, the action will be executed once the plugin is installed (in the `install` block) for each of the pipelines.
Hence, it will be only executed once before any call has ever arrived.